src/sentry/replays/lib/storage.py

```diff
@@ -0,0 +1,161 @@
+
+logger = logging.getLogger()
+
+
+@dataclasses.dataclass
+class RecordingSegmentStorageMeta:
+    project_id: int
+    replay_id: str
+    segment_id: int
+    retention_days: int
+    date_added: Optional[datetime] = None
+    file_id: Optional[int] = None
+
+
+class Blob(ABC):
+    @abstractmethod
+    def delete(self, segment: RecordingSegmentStorageMeta) -> None:
+        """Remove a blob from remote storage."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def get(self, segment: RecordingSegmentStorageMeta) -> bytes:
+        """Return blob from remote storage."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def set(self, segment: RecordingSegmentStorageMeta, value: bytes) -> None:
+        """Set blob in remote storage."""
+        raise NotImplementedError
+
+
+class FilestoreBlob(Blob):
+    """Filestore service driver blob manager."""
+
+    def delete(self, segment: RecordingSegmentStorageMeta) -> None:
+        file = File.objects.get(pk=segment.file_id)
+        file.delete()
+
+    @metrics.wraps("replays.lib.storage.FilestoreBlob.get")
+    def get(self, segment: RecordingSegmentStorageMeta) -> bytes:
+        file = File.objects.get(pk=segment.file_id)
+        return file.getfile().read()
+
+    @metrics.wraps("replays.lib.storage.FilestoreBlob.set")
+    def set(self, segment: RecordingSegmentStorageMeta, value: bytes) -> None:
+        file = File.objects.create(name=make_filename(segment), type="replay.recording")
+        file.putfile(BytesIO(value), blob_size=settings.SENTRY_ATTACHMENT_BLOB_SIZE)
+
+        try:
+            segment = ReplayRecordingSegment.objects.create(
+                project_id=segment.project_id,
+                replay_id=segment.replay_id,
+                segment_id=segment.segment_id,
+                file_id=file.id,
+                size=len(value),
+            )
+        except IntegrityError:
+            logger.warning(
+                "Recording-segment has already been processed.",
+                extra={
+                    "replay_id": segment.replay_id,
+                    "project_id": segment.project_id,
+                    "segment_id": segment.segment_id,
+                },
+            )
+
+            # Delete the file from remote storage since it is un-retrievable.
+            file.delete()
+
+
+class StorageBlob(Blob):
+    """Storage service driver blob manager.
+
+    This driver does not have managed TTLs.  To enable TTLs you will need to enable it on your
+    bucket.  Keys are prefixed by their TTL.  Those TTLs are 30, 60, 90.  Measured in days.
+    """
+
+    def delete(self, segment: RecordingSegmentStorageMeta) -> None:
+        storage = get_storage(self._make_storage_options())
+        storage.delete(self.make_key(segment))
+
+    @metrics.wraps("replays.lib.storage.StorageBlob.get")
+    def get(self, segment: RecordingSegmentStorageMeta) -> bytes:
+        cache_key = make_filename(segment)
```

Comment:
i wonder if it would be more optimal to cache at the endpoint level instead of here at the Storage, since we still have to decompress in the endpoint, i'm not sure if this will get us large gains.